packages/transaction-messages/src/compress-transaction-message.ts

```diff
@@ -0,0 +1,81 @@
+import { Address } from '@solana/addresses';
+import { AccountRole, IAccountLookupMeta, IInstruction, isSignerRole } from '@solana/instructions';
+
+import { AddressesByLookupTableAddress } from './addresses-by-lookup-table-address';
+import { TransactionMessage } from './transaction-message';
+
+type Mutable<T> = {
+    -readonly [P in keyof T]: T[P];
+};
+
+// Look up the address in lookup tables, return a lookup meta if it is found in any of them
+function findAddressInLookupTables(
+    address: Address,
+    role: AccountRole.READONLY | AccountRole.WRITABLE,
+    addressesByLookupTableAddress: AddressesByLookupTableAddress,
+): IAccountLookupMeta | undefined {
+    for (const [lookupTableAddress, addresses] of Object.entries(addressesByLookupTableAddress)) {
+        for (let i = 0; i < addresses.length; i++) {
+            if (address === addresses[i]) {
+                return {
+                    address,
+                    addressIndex: i,
+                    lookupTableAddress: lookupTableAddress as Address,
+                    role,
+                };
+            }
+        }
+    }
+}
+
+type TransactionMessageNotLegacy = Exclude<TransactionMessage, { version: 'legacy' }>;
+
+export function compressTransactionMessageUsingAddressLookupTables<
+    TTransactionMessage extends TransactionMessageNotLegacy = TransactionMessageNotLegacy,
+>(
+    transactionMessage: TTransactionMessage,
+    addressesByLookupTableAddress: AddressesByLookupTableAddress,
+): TTransactionMessage {
```

Comment:
This isn't exactly true, I guess. This function probably actually widens every account from whatever it is, back out to `IAccountMeta<TAddress> | IAccountLookupMeta<TAddress>`.

What would be the implications of mapping over this puppy and making that so?

```ts
// Pseudo TypeScript...
): 
  | TTransactionMessage  // Unchanged, or
  | TTransactionMessage & TTransactionMessage.instructions.map(i => ({
    ...i,
    accounts: i.accounts.map(a => IAccountMeta<typeof a.address> | IAccountLookupMeta<typeof a.address>)
  }))
{
```

Reply:
Haha I thought about this but the content of `AddressesByLookupTableAddress` would need to be fully static for this to even work which will probably never be the case.

Reply:
Fair point! I did this just to keep `ITransactionMessageWithFeePayer` etc intact on the output message, but it isn't strictly accurate if you've got the `accounts` field typed more specifically.

Reply:
Not 100% on this type but I think this works? https://github.com/solana-labs/solana-web3.js/pull/2969/commits/b643109329928112912a3bea5b6340f1135352a4